### PR TITLE
Deprecation policy - insert considerations for breaking changes

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
we have a number of structural changes ahead in Node Structure, fixture internals, as well as many other detail api's

as such i wanted to propose a update to our deprecation policy that enables sorting those out in a coordinated way when transitional apis are not sustainable.

### CODE REVIEW

File: doc/en/backwards-compatibility.rst [hunk 1]
```diff
@@ -3,6 +3,57 @@
 Backwards Compatibility Policy
 ==============================
 
+.. versionadded: 6.0
+
+pytest is actively being evolved and as a project that is been decades in the making,
+we keep learning about new and better structures to express different details about testing.
+
+While we implement those we try to ensure easy transition and don't want to impose unnecessary churn on our users and community/plugin authors.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
While we implement those modifications we try to ensure easy transition and don't want to impose unnecessary churn on our users and community/plugin authors.
```

File: doc/en/backwards-compatibility.rst [hunk 2]
```diff
@@ -3,6 +3,57 @@
...
+
+We are considering multipe Types of backward compatibility considering transitions
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
As of now, pytest considers multipe types of backward compatibility transitions:
```

After all we are not considering changing, we are changing correct?

File: doc/en/backwards-compatibility.rst [hunk 3]
```diff
@@ -3,6 +3,57 @@
...
+
+a) trivial: api's which trivially translate to the new mechanism,
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
**a) trivial**: APIs which trivially translate to the new mechanism,
```

File: doc/en/backwards-compatibility.rst [hunk 4]
```diff
@@ -3,6 +3,57 @@
...
+   and do not cause problematic details.
+
+   We try to support those indefinitely while encouraging users to switch to newer/better mechanisms.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   We try to support those indefinitely while encouraging users to switch to newer/better mechanisms through documentation.
```

File: doc/en/backwards-compatibility.rst [hunk 5]
```diff
@@ -3,6 +3,57 @@
...
+
+b) transitional: the old and new api don't conflict
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
**b) transitional**: the old and new API don't conflict
```

File: doc/en/backwards-compatibility.rst [hunk 6]
```diff
@@ -3,6 +3,57 @@
...
+   and we can help while supporting both for a prolonged time
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   and we can help users transition by using warnings, while supporting both for a prolonged time.
```

File: doc/en/backwards-compatibility.rst [hunk 7]
```diff
@@ -3,6 +3,57 @@
...
+
+   We will only start the removal of deprecated functionality in major releases (e.g. if we deprecate something in 3.0 we will start to remove it in 4.0), and keep it around for at least two minor releases (e.g. if we deprecate something in 3.9 and 4.0 is the next release, we start to remove it in 5.0, not in 4.0).
+
+   When the deprecation expires (e.g. 4.0 is released), we won't remove the deprecated functionality immediately, but will use the standard warning filters to turn them into **errors** by default. This approach makes it explicit that removal is imminent, and still gives you time to turn the deprecated feature into a warning instead of an error so it can be dealt with in your own time. In the next minor release (e.g. 4.1), the feature will be effectively removed.
+
+
+
+c) true breakage: should only to be cosidered when a combination of transition seems unreasonably unsustainable and would offset important development/features by years
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
**c) true breakage**: should only to be cosidered when normal transition is unreasonably unsustainable and would offset important development/features by years.
```

File: doc/en/backwards-compatibility.rst [hunk 8]
```diff
@@ -3,6 +3,57 @@
...
+   In addition they shuld be limited to apis where the number of actual users is very small and can be coordinated.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   In addition, they shuld be limited to APis where the number of actual users is very small (for example only impacting some plugins), and can be coordinated with the community in advance.
```

File: doc/en/backwards-compatibility.rst [hunk 9]
```diff
@@ -3,6 +3,57 @@
...
+
+   Examples for such upcinging changes:
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   Examples for such upcoming changes:
```

File: doc/en/backwards-compatibility.rst [hunk 10]
```diff
@@ -3,6 +3,57 @@
...
+
+   * removal of ``pytest_runtest_protocol/nextitem``
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]: Here we should either 1) explain why those should go and which impact they have or 2) link to an issue number

I understand some of them might not have an issue number yet (I think this one in particular doesn't), but as it is the examples are meaningless to users (and I suspect even some core devs), so they are not very useful.

File: doc/en/backwards-compatibility.rst [hunk 11]
```diff
@@ -3,6 +3,57 @@
...
+   * rearranging of the node tree to include FunctionDefinition
+   * rearranging of setupstate
+
+   True Breakages must be anounced first and only after coordination should be followed up by a initial POC Pull Request.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   True Breakages must be anounced first in an issue containing: 
   
   * Detailed description of the change
   * Rationale
   * Expected impact on users and plugin authors
   
   After there's no hard *-1* on the issue it should be followed up by a initial POC Pull Request.
   
```

File: doc/en/backwards-compatibility.rst [hunk 12]
```diff
@@ -3,6 +3,57 @@
...
+   This Proof of Concept serves as both, coordination point to assess Impact and potential inspriation to come up with a transitional solution after all.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   This Proof of Concept serves as both, coordination point to assess Impact and potential inspriation to come up with a transitional solution after all.
   
```

File: doc/en/backwards-compatibility.rst [hunk 13]
```diff
@@ -3,6 +3,57 @@
...
+   After a reasonable amount of time the PR can be merged to base a new Major release.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   After a reasonable amount of time the PR can be merged to base a new Major release.
   
```

File: doc/en/backwards-compatibility.rst [hunk 14]
```diff
@@ -3,6 +3,57 @@
...
+   The PR must contain a setup of DeprecationErrors that help fixing the related issues.
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   For the PR to mature from POC to acceptance, it must contain:
   * Setup of deprecation errors/warnings that help users fix and port their code. Note that if it is possible to introduce a deprecation period under the current series, before the true breakage, it should be introduced in a separate PR and be part of the current release stream.
   * Detailed description of the rationale and examples on how to port code in ``doc/en/deprecations.rst``.
```

Also I think this and the previous paragraph should be swapped.

File: doc/en/backwards-compatibility.rst [hunk 15]
```diff
@@ -3,6 +3,57 @@
...
+
+
+History
+=========
+
+
+
+
+focus primary on smoth transition - stance (pre 6.0)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by nicoddemus** [comment]
Do you plan to get this into 5.4 or only in the next major release?

File: doc/en/backwards-compatibility.rst [hunk 16]
```diff
@@ -3,6 +3,57 @@
 Backwards Compatibility Policy
 ==============================
 
+.. versionadded: 6.0
+
+pytest is actively being evolved and as a project that is been decades in the making,
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
pytest is actively being evolved and as a project that has been decades in the making,
```

File: doc/en/backwards-compatibility.rst (hunk 1 above)
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
While we implement those we try to ensure an easy transition and don't want to impose unnecessary churn on our users and community/plugin authors.
```

File: doc/en/backwards-compatibility.rst [hunk 17]
```diff
@@ -3,6 +3,57 @@
...
+we keep learning about new and better structures to express different details about testing.
+
+While we implement those we try to ensure easy transition and don't want to impose unnecessary churn on our users and community/plugin authors.
+
+We are considering multipe Types of backward compatibility considering transitions
+
+a) trivial: api's which trivially translate to the new mechanism,
+   and do not cause problematic details.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   and do not cause problematic changes.
```

File: doc/en/backwards-compatibility.rst (hunk 4 above)
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: Here and several times below, I don't think these indents render e.g. https://github.com/pytest-dev/pytest/blob/de33fd5d357120588a295b59828da6918148fe0a/doc/en/backwards-compatibility.rst

```suggestion
We try to support those indefinitely while encouraging users to switch to newer/better mechanisms.
```

File: doc/en/backwards-compatibility.rst [hunk 18]
```diff
@@ -3,6 +3,57 @@
...
+
+b) transitional: the old and new api don't conflict
+   and we can help while supporting both for a prolonged time
+
+   We will only start the removal of deprecated functionality in major releases (e.g. if we deprecate something in 3.0 we will start to remove it in 4.0), and keep it around for at least two minor releases (e.g. if we deprecate something in 3.9 and 4.0 is the next release, we start to remove it in 5.0, not in 4.0).
+
+   When the deprecation expires (e.g. 4.0 is released), we won't remove the deprecated functionality immediately, but will use the standard warning filters to turn them into **errors** by default. This approach makes it explicit that removal is imminent, and still gives you time to turn the deprecated feature into a warning instead of an error so it can be dealt with in your own time. In the next minor release (e.g. 4.1), the feature will be effectively removed.
+
+
+
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: Delete a couple of new lines

File: doc/en/backwards-compatibility.rst (hunk 8 above)
Comment (hugovk) [reply doc/en/backwards-compatibility.rst]: "APIs" :)

File: doc/en/backwards-compatibility.rst (hunk 11 above)
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: Spell out POC on first use, use POC for subsequent. 
```suggestion
   True breakages must be anounced first and only after coordination should be followed up by a initial proof-of-concept pull request.
```

File: doc/en/backwards-compatibility.rst (hunk 12 above)
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   This POC serves as both, coordination point to assess Impact and potential inspriation to come up with a transitional solution after all.
```

File: doc/en/backwards-compatibility.rst (hunk 13 above)
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   After a reasonable amount of time the PR can be merged to base a new major release.
```

File: doc/en/backwards-compatibility.rst [hunk 19]
```diff
@@ -3,6 +3,57 @@
...
+
+History
+=========
+
+
+
+
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: Can delete a bunch of newlines before and after History

File: doc/en/backwards-compatibility.rst [hunk 20]
```diff
@@ -3,6 +3,57 @@
...
+focus primary on smoth transition - stance (pre 6.0)
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
Focus primary on smooth transition - stance (pre 6.0)
```

File: doc/en/backwards-compatibility.rst (hunk 16 above)
Comment (asottile) [inline doc/en/backwards-compatibility.rst]:
```suggestion
pytest is actively evolving and is a project that has been decades in the making,
```

File: doc/en/backwards-compatibility.rst (hunk 2 above)
Comment (asottile) [reply doc/en/backwards-compatibility.rst]: "multiple" as well

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus i'd like to get this into 5.4, but enact it starting with 6.0

File: doc/en/backwards-compatibility.rst [hunk 21]
```diff
@@ -3,6 +3,68 @@
...
+
+As of now, pytest considers multipe types of backward compatibility transitions:
+
+**a) trivial**: APIs which trivially translate to the new mechanism,
+   and do not cause problematic changes.
+
+   We try to support those indefinitely while encouraging users to switch to newer/better mechanisms through documentation.
+
+**b) transitional**: the old and new API don't conflict
+   and we can help users transition by using warnings, while supporting both for a prolonged time.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: This renders a little strangely in GitHub:

image

https://github.com/RonnyPfannschmidt/pytest/blob/deprecation-breakage/doc/en/backwards-compatibility.rst#backwards-compatibility-policy

File: doc/en/backwards-compatibility.rst [hunk 22]
```diff
@@ -3,6 +3,68 @@
...
+
+   We will only start the removal of deprecated functionality in major releases (e.g. if we deprecate something in 3.0 we will start to remove it in 4.0), and keep it around for at least two minor releases (e.g. if we deprecate something in 3.9 and 4.0 is the next release, we start to remove it in 5.0, not in 4.0).
+
+   When the deprecation expires (e.g. 4.0 is released), we won't remove the deprecated functionality immediately, but will use the standard warning filters to turn them into **errors** by default. This approach makes it explicit that removal is imminent, and still gives you time to turn the deprecated feature into a warning instead of an error so it can be dealt with in your own time. In the next minor release (e.g. 4.1), the feature will be effectively removed.
+
+
+
+**c) true breakage**: should only to be cosidered when normal transition is unreasonably unsustainable and would offset important development/features by years.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
**c) true breakage**: should only to be considered when normal transition is unreasonably unsustainable and would delay important development/features by years.
```

File: doc/en/backwards-compatibility.rst [hunk 23]
```diff
@@ -3,6 +3,68 @@
...
+   In addition, they shuld be limited to APis where the number of actual users is very small (for example only impacting some plugins), and can be coordinated with the community in advance.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   In addition, they should be limited to APIs where the number of actual users is very small (for example only impacting some plugins), and can be coordinated with the community in advance.
```

File: doc/en/backwards-compatibility.rst [hunk 24]
```diff
@@ -3,6 +3,68 @@
...
+
+   Examples for such upcoming changes:
+
+   * removal of ``pytest_runtest_protocol/nextitem``
+   * rearranging of the node tree to include FunctionDefinition
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   * rearranging of the node tree to include ``FunctionDefinition``
```

File: doc/en/backwards-compatibility.rst [hunk 25]
```diff
@@ -3,6 +3,68 @@
...
+   * rearranging of setupstate
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: Is it this?
```suggestion
   * rearranging of ``SetupState``
```

File: doc/en/backwards-compatibility.rst [hunk 26]
```diff
@@ -3,6 +3,68 @@
...
+
+   True Breakages must be anounced first in an issue containing:
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   True breakages must be announced first in an issue containing: 
```

File: doc/en/backwards-compatibility.rst [hunk 27]
```diff
@@ -3,6 +3,68 @@
...
+   * rearranging of the node tree to include FunctionDefinition
+   * rearranging of setupstate
+
+   True Breakages must be anounced first in an issue containing: 
+   
+   * Detailed description of the change
+   * Rationale
+   * Expected impact on users and plugin authors
+   
+   After there's no hard *-1* on the issue it should be followed up by a initial POC Pull Request.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   After there's no hard *-1* on the issue it should be followed up by an initial proof-of-concept pull request.
```

File: doc/en/backwards-compatibility.rst [hunk 28]
```diff
@@ -3,6 +3,68 @@
...
+   
+   This POC serves as both, coordination point to assess Impact and potential inspriation to come up with a transitional solution after all.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
   This POC serves as both a coordination point to assess impact and potential inspiration to come up with a transitional solution after all.
```

File: doc/en/backwards-compatibility.rst [hunk 29]
```diff
@@ -3,6 +3,68 @@
...
+   
+   After a reasonable amount of time the PR can be merged to base a new major release.
+   
+   For the PR to mature from POC to acceptance, it must contain:
+   * Setup of deprecation errors/warnings that help users fix and port their code. Note that if it is possible to introduce a deprecation period under the current series, before the true breakage, it should be introduced in a separate PR and be part of the current release stream.
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]: "Note that" can be omitted for brevity:
```suggestion
   * Setup of deprecation errors/warnings that help users fix and port their code. If it is possible to introduce a deprecation period under the current series, before the true breakage, it should be introduced in a separate PR and be part of the current release stream.
```

File: doc/en/backwards-compatibility.rst [hunk 30]
```diff
@@ -3,6 +3,68 @@
...
+
+
+History
+=========
+
+
+
+
+Focus primary on smooth transition - stance (pre 6.0)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (hugovk) [inline doc/en/backwards-compatibility.rst]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: doc/en/backwards-compatibility.rst (hunk 4 above)
Comment (RonnyPfannschmidt) [reply doc/en/backwards-compatibility.rst]: just tried, those work fine

**Comment by RonnyPfannschmidt** [comment]
i'll squash before merging myself

File: doc/en/backwards-compatibility.rst [hunk 31]
```diff
@@ -3,6 +3,61 @@
...
+
+   * removal of ``pytest_runtest_protocol/nextitem``
+   * rearranging of the node tree to include ``FunctionDefinition``
+   * rearranging of ``SetupState``
+
+   True breakages must be announced first in an issue containing:
+
+   * Detailed description of the change
+   * Rationale
+   * Expected impact on users and plugin authors
```
Comment (nicoddemus) [inline doc/en/backwards-compatibility.rst]: We should link an existing issue that describes this type of breakage as an example (and is also an exercise to ensure the constructor API is following this protocol).

File: doc/en/backwards-compatibility.rst (hunk 4 above)
Comment (hugovk) [reply doc/en/backwards-compatibility.rst]: Rechecking myself, I agree they work fine. Sorry for the false alarm!

File: doc/en/backwards-compatibility.rst (hunk 31 above)
Comment (RonnyPfannschmidt) [reply doc/en/backwards-compatibility.rst]: followup in #6324

**Comment by RonnyPfannschmidt** [comment]
@asottile would you like another look, i'm going to squash soon and go for a merge then

**Comment by RonnyPfannschmidt** [comment]
thanks everyone for chiming in and helping out to make this one nice :+1:

**Comment by nicoddemus** [comment]
I think we should announce this somehow, would you like to post this on Twitter @RonnyPfannschmidt?

**Comment by nicoddemus** [comment]
I just realized this went to `features`... shouldn't we publish this right away?

**Comment by RonnyPfannschmidt** [comment]
It's intentionally there, the next and possible last feature branch feature release is close

**Comment by nicoddemus** [comment]
Cool.

Btw can your review your action items for [6.0](https://github.com/pytest-dev/pytest/milestone/24)? I suspect the ones there created by you won't be ready for 6.0 and will need to be postponed.

**Comment by RonnyPfannschmidt** [comment]
Will take a look after travel

**Pull request merged**